### PR TITLE
codeintel: Add tracing to bundle manager client

### PR DESCRIPTION
**Pull request by efritz**
Add tracing around bundle manager client.

### CODE REVIEW

File: internal/codeintel/bundles/client/bundle_manager_client.go [hunk 1]
```diff
@@ -38,15 +45,44 @@ type baseClient interface {
...
 }
 
 var _ BundleManagerClient = &bundleManagerClientImpl{}
 var _ baseClient = &bundleManagerClientImpl{}
 
 func New(bundleManagerURL string) BundleManagerClient {
-	return &bundleManagerClientImpl{bundleManagerURL: bundleManagerURL}
+	return &bundleManagerClientImpl{
+		httpClient:       &http.Client{Transport: defaultTransport},
+		httpLimiter:      parallel.NewRun(500),
```
Comment (gbrik) [inline internal/codeintel/bundles/client/bundle_manager_client.go:82]: nit:
make the 500 a constant

File: internal/codeintel/bundles/client/bundle_manager_client.go [hunk 2]
```diff
@@ -38,15 +45,44 @@ type baseClient interface {
...
+	}, func(u *url.URL) string {
+		// Extract the operation from a path like `/dbs/{id}/{operation}`
+		if segments := strings.Split(u.Path, "/"); len(segments) == 4 {
+			return segments[3]
+		}
+
+		// All other methods are uploading/downloading upload or bundle files.
+		// These can all go into a single bucket which  are meant not necessarily
+		// meant to have low latency.
+		return "transfer"
```
Comment (gbrik) [inline internal/codeintel/bundles/client/bundle_manager_client.go:65]: nit:
this is brittle

File: monitoring/precise_code_intel_api_server.go [hunk 3]
```diff
@@ -73,6 +71,55 @@ func PreciseCodeIntelAPIServer() *Container {
...
+							PossibleSolutions: "none",
+						},
+						{
+							Name:              "99th_percentile_bundle_manager_duration",
+							Description:       "99th percentile successful bundle manager data transfer duration over 5m",
+							Query:             `histogram_quantile(0.99, sum by (le,category)(rate(src_precise_code_intel_bundle_manager_request_duration_seconds_bucket{job="precise-code-intel-api-server",category="transfer"}[5m])))`,
+							DataMayNotExist:   true,
+							Warning:           Alert{GreaterOrEqual: 300},
+							PanelOptions:      PanelOptions().LegendFormat("{{category}}").Unit(Seconds),
+							PossibleSolutions: "none",
```
Comment (gbrik) [inline monitoring/precise_code_intel_api_server.go:91]: dup'd names

Comment (efritz) [reply monitoring/precise_code_intel_api_server.go:91]: One is query and one is data transfer.

Comment (efritz) [reply monitoring/precise_code_intel_api_server.go]: Nvm, got it.

File: internal/codeintel/bundles/client/bundle_manager_client.go (hunk 2 above)
Comment (efritz) [reply internal/codeintel/bundles/client/bundle_manager_client.go:65]: What alternative do you suggest?

Comment (efritz) [reply internal/codeintel/bundles/client/bundle_manager_client.go:67]: This is also a common pattern in our codebase (bitbucketcloud does the same split logic, as does extsvc github, gitlab, as does search).

Please open an issue if you have a better way to organize this (or are sufficiently nitt'd).

Comment (efritz) [reply internal/codeintel/bundles/client/bundle_manager_client.go:67]: Ah ok. I'm leaving what's in there as-is for now as we won't really be expanding the routes very frequently. I also wanted to be able to use the request method to differentiate between upload and download, but the function signature doesn't provide that.

I agree that a statically checked set of routes would be the way to go here. We should update all of the places where this method is used at the same time.

**Pull request merged**